src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7848,8 +7873,10 @@ private ForEachStatementSyntax ParseForEachStatement()
             // Can be a 'for' keyword if the user typed: 'for (SomeType t in'
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ForEachKeyword || this.CurrentToken.Kind == SyntaxKind.ForKeyword);
 
-            // Syntax for foreach is:
+            // Syntax for foreach is either:
             //  foreach ( <type> <identifier> in <expr> ) <embedded-statement>
+            // or
+            //  foreach ( <deconstruction-variables> in <expr> ) <embedded-statement>
```

Comment:
The name implies to me that it can only be variables.  I assume we can have declarations here as well.